File: scripts/monitoring/watcher.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'fileutils'
require 'time'
require 'optparse'

module Gmux
  module Scripts
    class Watcher
      STORE_PATH = File.join(Dir.home, '.gmux', 'sessions.json')
      LOG_DIR = File.join(Dir.home, '.gmux', 'logs')

      ERROR_PATTERNS = [
        /error/i,
        /exception/i,
        /failed/i,
        /fatal/i,
        /panic/i,
        /crash/i,
        /killed/i,
        /segfault/i,
        /errno/i,
        /EACCES/i,
        /ENOENT/i,
        /ENOMEM/i,
        /timeout/i,
        /timed out/i,
        /connection refused/i,
        /permission denied/i,
        /no such file/i,
        /command not found/i,
        /syntax error/i,
        /undefined symbol/i,
        /module not found/i,
        /import error/i,
        /cannot find/i,
        /not found/i,
        /denied/i,
        /forbidden/i,
      ].freeze

      def initialize(options = {})
        @interval = options[:interval] || 5
        @verbose = options[:verbose] || false
        @log_file = options[:log_file]
        @patterns = options[:patterns] || []
        @watch_session = options[:session]
        @tail = options[:tail] || false
      end

      def run
        puts "=== GMUX Agent Watcher ==="
        puts "Interval: #{@interval}s"
        puts "Patterns: #{@patterns.length > 0 ? @patterns.join(', ') : 'default error patterns'}"
        puts

        ensure_log_dir

        if @tail
          tail_output
        else
          watch_sessions
        end
      end

      private

      def ensure_log_dir
        FileUtils.mkdir_p(LOG_DIR)
      end

      def load_sessions
        return {} unless File.exist?(STORE_PATH)
        JSON.parse(File.read(STORE_PATH))
      rescue JSON::ParserError
        {}
      end

      def watch_sessions
        puts "Watching for errors... (Ctrl+C to stop)"
        puts

        running = true
        trap("INT") { running = false }

        while running
          sessions = load_sessions
          sessions_to_watch = @watch_session ?
            sessions.select { |name, _| name == @watch_session } :
            sessions.select { |_, s| s['status'] == 'running' }

          if sessions_to_watch.empty?
            print "\r  No running sessions to watch..."
            sleep @interval
            next
          end

          sessions_to_watch.each do |name, session|
            watch_session(name, session)
          end

          sleep @interval
        end

        puts "\nStopped watching."
      end

      def watch_session(name, session)
        pane_id = session['tmuxPaneId']
        return unless pane_id

        # Get recent output from tmux pane
        output = capture_pane_output(pane_id)
        return if output.empty?

        # Check for errors
        errors = detect_errors(output)
        if errors.any?
          log_errors(name, errors)
          display_errors(name, errors)
        end
      end

      def capture_pane_output(pane_id)
        # Capture last 100 lines from tmux pane
        result = `tmux capture-pane -t #{pane_id} -p -S -100 2>/dev/null`
        $?.success? ? result : ''
      end

      def detect_errors(output)
        errors = []
        lines = output.split("\n")

        lines.each_with_index do |line, i|
          # Skip empty lines
          next if line.strip.empty?

          # Check against patterns
          patterns_to_check = @patterns.empty? ? ERROR_PATTERNS : @patterns.map { |p| Regexp.new(p) }

          patterns_to_check.each do |pattern|
            if line.match?(pattern)
              errors << {
                line: i + 1,
                content: line.strip,
                pattern: pattern.source
              }
              break
            end
          end
        end

        errors.uniq { |e| e[:content] }
      end

      def log_errors(session_name, errors)
        timestamp = Time.now.strftime('%Y-%m-%d_%H-%M-%S')
        log_file = @log_file || File.join(LOG_DIR, "#{session_name}_errors.log")

        File.open(log_file, 'a') do |f|
          f.puts "=" * 60
          f.puts "Session: #{session_name}"
          f.puts "Time: #{Time.now.iso8601}"
          f.puts "Errors found: #{errors.length}"
          f.puts "-" * 60
          errors.each do |error|
            f.puts "Line #{error[:line]}: #{error[:content]}"
          end
          f.puts
        end

        puts "  Logged #{errors.length} error(s) to: #{log_file}" if @verbose
      end

      def display_errors(session_name, errors)
        puts
        puts "  [!] #{session_name}: #{errors.length} error(s) detected"
        errors.first(5).each do |error|
          puts "      Line #{error[:line]}: #{error[:content][0..80]}"
        end
        puts "      ... and #{errors.length - 5} more" if errors.length > 5
      end

      def tail_output
        puts "Tailing output... (Ctrl+C to stop)"
        puts

        running = true
        trap("INT") { running = false }

        sessions = load_sessions
        sessions_to_watch = @watch_session ?
          sessions.select { |name, _| name == @watch_session } :
          sessions.select { |_, s| s['status'] == 'running' }

        if sessions_to_watch.empty?
          puts "No running sessions to tail."
          return
        end

        pane_ids = sessions_to_watch.map { |_, s| s['tmuxPaneId'] }.compact

        # Use tmux pipe-pane for real-time output
        pane_ids.each do |pane_id|
          puts "Tailing pane: #{pane_id}"
          system("tmux pipe-pane -t #{pane_id} 'cat >> #{LOG_DIR}/pane_#{pane_id}.log'")
        end

        # Monitor log files
        while running
          pane_ids.each do |pane_id|
            log_file = File.join(LOG_DIR, "pane_#{pane_id}.log")
            next unless File.exist?(log_file)

            # Check for new content
            size = File.size(log_file)
            if size > @last_sizes[pane_id]
              content = File.read(log_file, offset: @last_sizes[pane_id])
              print content
              @last_sizes[pane_id] = size
            end
          end
          sleep 0.1
        end

        # Stop pipe-pane
        pane_ids.each do |pane_id|
          system("tmux pipe-pane -t #{pane_id} -c")
        end

        puts "\nStopped tailing."
      end
    end
  end
end

if __FILE__ == $PROGRAM_NAME
  options = { last_sizes: {} }
  OptionParser.new do |opts|
    opts.banner = "Usage: #{$PROGRAM_NAME} [options]"

    opts.on('-i', '--interval SECONDS', Integer, "Check interval in seconds (default: 5)") do |i|
      options[:interval] = i
    end

    opts.on('-s', '--session NAME', "Watch specific session") do |s|
      options[:session] = s
    end

    opts.on('-p', '--pattern REGEX', Array, "Custom error patterns (regex)") do |p|
      options[:patterns] = p
    end

    opts.on('-l', '--log-file FILE', "Log errors to file") do |l|
      options[:log_file] = l
    end

    opts.on('-t', '--tail', "Tail pane output in real-time") do
      options[:tail] = true
    end

    opts.on('-v', '--verbose', "Show detailed output") do
      options[:verbose] = true
    end

    opts.on('-h', '--help', "Show this help") do
      puts opts
      exit
    end
  end.parse!

  Gmux::Scripts::Watcher.new(options).run
end

File: scripts/monitoring/watcher.py
#!/usr/bin/env python3
"""GMUX Agent Watcher - Monitor agent output for errors."""

import json
import logging
import os
import re
import signal
import subprocess
import sys
import time
from argparse import ArgumentParser
from datetime import datetime
from pathlib import Path

# Setup logging
logging.basicConfig(
    level=logging.INFO,
    format="%(message)s",
)
logger = logging.getLogger(__name__)

# Error patterns to detect
ERROR_PATTERNS = [
    r"error",
    r"exception",
    r"failed",
    r"fatal",
    r"panic",
    r"crash",
    r"killed",
    r"segfault",
    r"errno",
    r"EACCES",
    r"ENOENT",
    r"ENOMEM",
    r"timeout",
    r"timed out",
    r"connection refused",
    r"permission denied",
    r"no such file",
    r"command not found",
    r"syntax error",
    r"undefined symbol",
    r"module not found",
    r"import error",
    r"cannot find",
    r"not found",
    r"denied",
    r"forbidden",
]

STORE_PATH = Path.home() / ".gmux" / "sessions.json"
LOG_DIR = Path.home() / ".gmux" / "logs"


class Watcher:
    def __init__(self, options):
        self.interval = options.get("interval", 5)
        self.verbose = options.get("verbose", False)
        self.log_file = options.get("log_file")
        self.patterns = options.get("patterns", [])
        self.watch_session = options.get("session")
        self.tail = options.get("tail", False)
        self.running = True
        self.last_sizes = {}

        # Compile patterns
        self.compiled_patterns = [
            re.compile(p, re.IGNORECASE)
            for p in (self.patterns or ERROR_PATTERNS)
        ]

        # Setup signal handler
        signal.signal(signal.SIGINT, self._handle_interrupt)

    def _handle_interrupt(self, signum, frame):
        self.running = False

    def run(self):
        print("=== GMUX Agent Watcher ===")
        print(f"Interval: {self.interval}s")
        patterns_desc = (
            ", ".join(self.patterns) if self.patterns else "default error patterns"
        )
        print(f"Patterns: {patterns_desc}")
        print()

        self.ensure_log_dir()

        if self.tail:
            self.tail_output()
        else:
            self.watch_sessions()

    def ensure_log_dir(self):
        LOG_DIR.mkdir(parents=True, exist_ok=True)

    def load_sessions(self):
        if not STORE_PATH.exists():
            return {}
        try:
            return json.loads(STORE_PATH.read_text())
        except (json.JSONDecodeError, OSError):
            return {}

    def watch_sessions(self):
        print("Watching for errors... (Ctrl+C to stop)")
        print()

        while self.running:
            sessions = self.load_sessions()
            sessions_to_watch = (
                {k: v for k, v in sessions.items() if k == self.watch_session}
                if self.watch_session
                else {k: v for k, v in sessions.items() if v.get("status") == "running"}
            )

            if not sessions_to_watch:
                print("\r  No running sessions to watch...", end="", flush=True)
                time.sleep(self.interval)
                continue

            for name, session in sessions_to_watch.items():
                self._watch_session(name, session)

            time.sleep(self.interval)

        print("\nStopped watching.")

    def _watch_session(self, name, session):
        pane_id = session.get("tmuxPaneId")
        if not pane_id:
            return

        output = self._capture_pane_output(pane_id)
        if not output:
            return

        errors = self._detect_errors(output)
        if errors:
            self._log_errors(name, errors)
            self._display_errors(name, errors)

    def _capture_pane_output(self, pane_id):
        try:
            result = subprocess.run(
                ["tmux", "capture-pane", "-t", pane_id, "-p", "-S", "-100"],
                capture_output=True,
                text=True,
                timeout=5,
            )
            return result.stdout if result.returncode == 0 else ""
        except (subprocess.TimeoutExpired, FileNotFoundError):
            return ""

    def _detect_errors(self, output):
        errors = []
        seen = set()

        for i, line in enumerate(output.split("\n"), 1):
            if not line.strip():
                continue

            for pattern in self.compiled_patterns:
                if pattern.search(line):
                    error_content = line.strip()
                    if error_content not in seen:
                        errors.append(
                            {
                                "line": i,
                                "content": error_content,
                                "pattern": pattern.pattern,
                            }
                        )
                        seen.add(error_content)
                    break

        return errors

    def _log_errors(self, session_name, errors):
        log_file = self.log_file or (LOG_DIR / f"{session_name}_errors.log")

        with open(log_file, "a") as f:
            f.write("=" * 60 + "\n")
            f.write(f"Session: {session_name}\n")
            f.write(f"Time: {datetime.now().isoformat()}\n")
            f.write(f"Errors found: {len(errors)}\n")
            f.write("-" * 60 + "\n")
            for error in errors:
                f.write(f"Line {error['line']}: {error['content']}\n")
            f.write("\n")

        if self.verbose:
            print(f"  Logged {len(errors)} error(s) to: {log_file}")

    def _display_errors(self, session_name, errors):
        print()
        print(f"  [!] {session_name}: {len(errors)} error(s) detected")
        for error in errors[:5]:
            content = error["content"][:80]
            print(f"      Line {error['line']}: {content}")
        if len(errors) > 5:
            print(f"      ... and {len(errors) - 5} more")

    def tail_output(self):
        print("Tailing output... (Ctrl+C to stop)")
        print()

        sessions = self.load_sessions()
        sessions_to_watch = (
            {k: v for k, v in sessions.items() if k == self.watch_session}
            if self.watch_session
            else {k: v for k, v in sessions.items() if v.get("status") == "running"}
        )

        if not sessions_to_watch:
            print("No running sessions to tail.")
            return

        pane_ids = [s.get("tmuxPaneId") for s in sessions_to_watch.values() if s.get("tmuxPaneId")]

        # Start pipe-pane for each pane
        for pane_id in pane_ids:
            log_file = LOG_DIR / f"pane_{pane_id}.log"
            try:
                subprocess.run(
                    ["tmux", "pipe-pane", "-t", pane_id, f"cat >> {log_file}"],
                    check=False,
                )
            except FileNotFoundError:
                pass

        # Monitor log files
        while self.running:
            for pane_id in pane_ids:
                log_file = LOG_DIR / f"pane_{pane_id}.log"
                if not log_file.exists():
                    continue

                size = log_file.stat().st_size
                last_size = self.last_sizes.get(pane_id, 0)

                if size > last_size:
                    with open(log_file, "r") as f:
                        f.seek(last_size)
                        content = f.read()
                        print(content, end="")
                    self.last_sizes[pane_id] = size

            time.sleep(0.1)

        # Stop pipe-pane for each pane
        for pane_id in pane_ids:
            try:
                subprocess.run(
                    ["tmux", "pipe-pane", "-t", pane_id, "-c"],
                    check=False,
                )
            except FileNotFoundError:
                pass

        print("\nStopped tailing.")


def main():
    parser = ArgumentParser(description="Monitor GMUX agent output for errors")
    parser.add_argument(
        "-i",
        "--interval",
        type=int,
        default=5,
        help="Check interval in seconds (default: 5)",
    )
    parser.add_argument(
        "-s",
        "--session",
        help="Watch specific session",
    )
    parser.add_argument(
        "-p",
        "--pattern",
        action="append",
        dest="patterns",
        help="Custom error patterns (regex)",
    )
    parser.add_argument(
        "-l",
        "--log-file",
        dest="log_file",
        help="Log errors to file",
    )
    parser.add_argument(
        "-t",
        "--tail",
        action="store_true",
        help="Tail pane output in real-time",
    )
    parser.add_argument(
        "-v",
        "--verbose",
        action="store_true",
        help="Show detailed output",
    )

    args = parser.parse_args()

    watcher = Watcher(vars(args))
    watcher.run()


if __name__ == "__main__":
    main()
